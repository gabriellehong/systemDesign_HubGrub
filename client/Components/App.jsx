import React from 'react';
import ReviewHeader from './ReviewHeader/ReviewHeader.jsx';
import ReviewList from './ReviewList/ReviewList.jsx';
import Footer from './Footer/Footer.jsx';
import axios from 'axios';
import restaurantData from '../../restaurantData.js';

// let axios_url = (process.env.NODE_ENV === 'development') ? 'http://localhost:3200' : 'https://reviews-footer.herokuapp.com'
let axios_url = 'http://localhost:3200'

class App extends React.Component {
  constructor() {
    super();
    this.state = {
      reviewData: [], //all reviews from restaurant
      restaurantData: {} //data of restaurant
    }
    this.getReviewsForRestaurantID = this.getReviewsForRestaurantID.bind(this);
    this.getRestaurantDataByRestaurantID = this.getRestaurantDataByRestaurantID.bind(this);
    this.sortReviews = this.sortReviews.bind(this);
  }

  componentDidMount() {
    this.getReviewsForRestaurantID();
  }

  getReviewsForRestaurantID() {
    let id = window.location.href.split("id=")[1];
    // console.log(id); //undefined
    // let url = window.location.href;
    // let {id} = this.state
    axios.get(`${axios_url}/restaurants/psql/reviews/${id}`)
      .then( data => {
        this.sortReviews('Most recent', data.data);
      }, (data) => console.log(data.data))
      .catch(err => console.log('Error getting review data: ',err));
  }
  
  getRestaurantDataByRestaurantID(data) {
    //get the restaurant id by looking at the restaurantID of the first review (assuming there is at least one review)
    if (data.length > 0) {
      let id = data[0].restaurantID;
      //loop over restaurant data to find the restaurant that matches the id
      for (let i = 0; i < restaurantData.length; i++) {
        if (id === restaurantData[i].id) {
          return restaurantData[i];
        }
      }
    } else {
      return {};
    }
  }

  sortReviews(method, data = this.state.reviewData) {
    if (method === 'Most recent') {
      let sortedReviewData = data.sort(function(a,b){
        return new Date(b.date) - new Date(a.date);
      })
      let restaurantData = this.getRestaurantDataByRestaurantID(sortedReviewData);
      this.setState({reviewData: sortedReviewData, restaurantData: restaurantData})

    } else if (method === 'Highest rated') {
      //sort this.state.reviewData by starRating
      let sortedReviewData = data.sort(function(a,b){
        return b.starRating - a.starRating;
      })
      this.setState({reviewData: sortedReviewData})
    }
  }

  render() {
    return (
      <div>
        <ReviewHeader restaurantData={this.state.restaurantData}/>
        <ReviewList reviewData={this.state.reviewData} sortReviews={this.sortReviews} restaurantData={this.state.restaurantData}/>
        <Footer />
      </div>
    )
  }
}

export default App;