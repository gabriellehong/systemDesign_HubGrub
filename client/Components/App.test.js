import { shallow, mount, render } from 'enzyme';
import renderer from 'react-test-renderer';
import React from 'react';
import App from './App';

describe('testing App component', ()=> {

	it('should render App component', () => {
		expect(shallow(<App />).length).toEqual(1)
	})

	it('should grab a snapshot of the component', ()=> {
		const tree = renderer
		  .create(<App />)
		  .toJSON();
		expect(tree).toMatchSnapshot();
	})

	it('should have default state as empty objects', ()=> {
		let wrapper = shallow(<App />);
		expect.assertions(2);
		expect(wrapper.state().reviewData).toEqual([]);
		expect(wrapper.state().restaurantData).toEqual({});
	})

	it('should render componentDidMount()', ()=> {
		let wrapper = mount(<App />);
		let instance = wrapper.instance();
		jest.spyOn(instance, 'getReviewsForRestaurantID');
		instance.componentDidMount();
		expect(instance.getReviewsForRestaurantID).toHaveBeenCalledTimes(1);
	})

	it('should sort by most recent', ()=> {
		let wrapper = mount(<App />);
		wrapper.instance().sortReviews('Most recent', [])
		expect(wrapper.state().reviewData).toEqual([]);
	})

	it('should sort by highest rated', ()=> {
		let wrapper = mount(<App />);
		wrapper.instance().sortReviews('Highest rated', [])
		expect(wrapper.state().reviewData).toEqual([]);
	})

	it('should sort by highest rated', ()=> {
		let wrapper = mount(<App />);
		let result = wrapper.instance().getRestaurantDataByRestaurantID([{"_id":"5ce474fb7413717579ce211e","Ordered":[],"user":{"firstName":"Silas","firstLetter":"S","numOfRatings":10,"topReviewer":false},"reviewID":1,"restaurantID":2,"starRating":1,"comments":"I hate this restaurant. I usually love Kimchi but they don't know how to make it the right way. I would not feed this to my dog.","date":"2018-09-05T09:11:40.594Z","__v":0}])
		expect(result).toEqual({ 
			id: 2,
			name: 'Lee\'s Kitchen',
			address: { 
				line1: '1893 E Tanner St',
				line2: 'Wileyburgh, LA 53498',
				milesAway: 1.56
			},
			phone: '(856) 264-0286',
			ratings: { 
				number: 466, 
				stars: 5, 
				good: 99, 
				timely: 91, 
				correct: 97 
			},
			schedule: {
				monFri: { delivery: '11:00am-8:30pm', pickup: '10:30am-10:30pm' },
				satSun: { delivery: 'Closed', pickup: 'Closed' }
			},
			image: 'https://s3-us-west-1.amazonaws.com/kayjayhogan/Korean/chicken-soup-1346310_1280.jpg',
			priceRange: 3,
			categories: [ 'Korean', 'Dinner', 'Soup', 'Salads' ] 
		});
	})

})