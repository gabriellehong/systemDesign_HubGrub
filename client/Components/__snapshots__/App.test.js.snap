// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`testing App component should grab a snapshot of the component 1`] = `
<div>
  <div
    className="reviewHeader-main"
  >
    <h3
      className="reviewHeader-title"
    >
      Reviews for 
    </h3>
    <div />
    <div
      className="reviewHeader-numOfRatings"
    >
       ratings
    </div>
    <div
      className="reviewHeader-stats"
    >
      <div
        className="reviewHeader-statsTitle"
      >
        Here's what people are saying:
      </div>
      <ul
        className="reviewHeader-ul"
      >
        <li
          className="reviewHeader-li"
        >
          <span
            className="reviewHeader-ratings"
          >
            %
          </span>
          <span
            className="reviewHeader-caption"
          >
            Food was good
          </span>
        </li>
        <li
          className="reviewHeader-li"
        >
          <span
            className="reviewHeader-ratings"
          >
            %
          </span>
          <span
            className="reviewHeader-caption"
          >
            Delivery was on time
          </span>
        </li>
        <li
          className="reviewHeader-li"
        >
          <span
            className="reviewHeader-ratings"
          >
            %
          </span>
          <span
            className="reviewHeader-caption"
          >
            Order was accurate
          </span>
        </li>
      </ul>
    </div>
  </div>
  <div>
    <div
      className="reviewList-main"
    >
      <div
        className="reviewList-sort"
      >
        <div
          className="reviewList-sortText"
        >
          Sort by:
        </div>
        <div
          className="reviewList-select"
        >
          <select
            className="reviewList-selectBox"
            onChange={[Function]}
          >
            <option
              value="Most recent"
            >
              Most recent
            </option>
            <option
              value="Highest rated"
            >
              Highest rated
            </option>
          </select>
          <img
            alt="dropdownArrow"
            className="reviewList-dropdownArrow"
            src="https://s3-us-west-1.amazonaws.com/fechackreactor/grubhubDropdownArrow.png"
          />
        </div>
      </div>
      <br />
      <div
        className="reviewList-buttonContainer"
      >
        <button
          className="reviewList-viewAllButton"
          id="reviewList-viewAllButton"
          onClick={[Function]}
        >
          View all reviews
        </button>
      </div>
      <div
        className="reviewList-bottom"
      >
        <a
          className="reviewList-a"
          href=""
        >
          Grubhub
        </a>
         /
        <a
          className="reviewList-a"
          href=""
        >
          Los Angeles
        </a>
         /
        <a
          className="reviewList-a"
          href=""
        >
          Santa Monica
        </a>
         /
        <span
          className="reviewList-restaurantName"
        />
      </div>
    </div>
  </div>
  <div
    className="footer-body"
  >
    <div
      className="footer-topHalf"
    >
      <div
        className="footer-collapsibleSection col-1-4"
      >
        <span
          className="footer-collapsibletitleText"
        >
          Get to know us
        </span>
        <div
          className="footer-collapsibleContent"
        >
          <ul>
            <li
              className="footer-li"
            >
              <a
                className="footer-a"
                href=""
              >
                About Grubhub
              </a>
            </li>
            <li
              className="footer-li"
            >
              <a
                className="footer-a"
                href=""
              >
                Our apps
              </a>
            </li>
            <li
              className="footer-li"
            >
              <a
                className="footer-a"
                href=""
              >
                Our blog
              </a>
            </li>
            <li
              className="footer-li"
            >
              <a
                className="footer-a"
                href=""
              >
                Careers
              </a>
            </li>
            <li
              className="footer-li"
            >
              <a
                className="footer-a"
                href=""
              >
                Investor relations
              </a>
            </li>
            <li
              className="footer-li"
            >
              <a
                className="footer-a"
                href=""
              >
                News
              </a>
            </li>
          </ul>
        </div>
      </div>
      <div
        className="footer-collapsibleSection col-1-4"
      >
        <span
          className="footer-collapsibletitleText"
        >
          Useful links
        </span>
        <div
          className="footer-collapsibleContent"
        >
          <ul
            className="footer-ul"
          >
            <li
              className="footer-li"
            >
              <a
                className="footer-a"
                href=""
              >
                Gift cards
              </a>
            </li>
            <li
              className="footer-li"
            >
              <a
                className="footer-a"
                href=""
              >
                Coupons
              </a>
            </li>
            <li
              className="footer-li"
            >
              <a
                className="footer-a"
                href=""
              >
                FAQ
              </a>
            </li>
            <li
              className="footer-li"
            >
              <a
                className="footer-a"
                href=""
              >
                Help
              </a>
            </li>
            <li
              className="footer-li"
            >
              <a
                className="footer-a"
                href=""
              >
                Grubhub for Alexa
              </a>
            </li>
            <li
              className="footer-li"
            >
              <a
                className="footer-a"
                href=""
              >
                Student discounts
              </a>
            </li>
          </ul>
        </div>
      </div>
      <div
        className="footer-collapsibleSection col-1-4"
      >
        <span
          className="footer-collapsibletitleText"
        >
          Connect with us
        </span>
        <div
          className="footer-collapsibleContent"
        >
          <ul>
            <li
              className="footer-li"
            >
              <a
                className="footer-a"
                href=""
              >
                Facebook
              </a>
            </li>
            <li
              className="footer-li"
            >
              <a
                className="footer-a"
                href=""
              >
                Twitter
              </a>
            </li>
            <li
              className="footer-li"
            >
              <a
                className="footer-a"
                href=""
              >
                Instagram
              </a>
            </li>
            <li
              className="footer-li"
            >
              <a
                className="footer-a"
                href=""
              >
                YouTube
              </a>
            </li>
          </ul>
        </div>
      </div>
      <div
        className="footer-collapsibleSection col-1-4"
      >
        <span
          className="footer-collapsibletitleText"
        >
          Partner with us
        </span>
        <div
          className="footer-collapsibleContent"
        >
          <ul>
            <li
              className="footer-li"
            >
              <a
                className="footer-a"
                href=""
              >
                For restaurants
              </a>
            </li>
            <li
              className="footer-li"
            >
              <a
                className="footer-a"
                href=""
              >
                For drivers
              </a>
            </li>
            <li
              className="footer-li"
            >
              <a
                className="footer-a"
                href=""
              >
                For corporate accounts
              </a>
            </li>
          </ul>
        </div>
      </div>
    </div>
    <div
      className="footer-bottomHalf"
    >
      <div
        className="footer-copyright"
      >
        Â© 2019 Grubhub All rights reserved.
      </div>
      <a
        className="footer-a footer-terms"
        href=""
      >
        Terms of Use
      </a>
      <br />
      <a
        className="footer-a footer-privacy"
        href=""
      >
        Privacy Policy
      </a>
    </div>
  </div>
</div>
`;
